subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java

```diff
@@ -160,6 +161,32 @@ public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI ur
         }
     }
 
+    private void checkInstallFolderForLeftoverContent(File installFolder, URI uri, JavaToolchainSpec spec, JvmInstallationMetadata metadata) {
+        if (!installFolder.exists()) {
+            return; //install folder doesn't even exist
+        }
+
+        File[] filesInInstallFolder = installFolder.listFiles();
+        if (filesInInstallFolder == null || filesInInstallFolder.length == 0) {
+            return; //no files in install folder
+        }
+
+        File markerLocation = markedLocation(installFolder);
+        if (!isMarkedLocation(markerLocation)) {
+            return; //no marker found
+        }
+
+        String leftoverMetadata;
+        try {
+            leftoverMetadata = getMetadata(markerLocation).toString();
+        } catch (Exception e) {
+            leftoverMetadata = "Could not be determined due to: " + e.getMessage();
+        }
+        LOGGER.warn("While provisioning Java toolchain from '{}' to satisfy spec '{}', having metadata '{}', " +
+                "leftover content with metadata '{}' found in install folder '{}'. Existing installation will be replaced by the new download.",
```

Comment:
Suggestion for better flow of the message:
```suggestion
        LOGGER.warn("While provisioning Java toolchain from '{}' to satisfy spec '{}' (with metadata '{}') " +
                "leftover content (with metadata '{}') was found in the install folder '{}'. The existing installation will be replaced by the new download.",
```

Reply:
Feedback from native English speakers always welcome.